File: Codejam/2020/KickStart/RoundA/workout.cpp
#include <bits/stdc++.h>

using namespace std;
const int MAX_N = 100005;
int numTest;
int a[MAX_N];
int n, K;

int main () {
  freopen("input.txt", "r", stdin);
  freopen("output.txt", "w", stdout);
  cin >> numTest;
  for (int testCase = 1; testCase <= numTest; testCase++) {
    cout << "Case #" << testCase << ": ";
    cin >> n >> K;
    for (int i = 1; i <= n; i++) {
      cin >> a[i];
    }
    int low = 1;
    int high = 1000000007;
    int ans = 0;
    while (low <= high) {
      int mid = (low + high) >> 1;
      int numAdd = 0;
      for (int i = 2; i <= n; i++) {
        int diff = a[i] - a[i - 1];
        if (diff <= mid) {
          continue;
        }
        numAdd += diff / mid;
        if (diff % mid == 0) {
          numAdd--;
        }
      }
      if (numAdd <= K) {
        ans = mid;
        high = mid - 1;
      }
      else {
        low = mid + 1;
      }
    }
    cout << ans << endl;
  }
  return 0;
}

File: Codejam/2020/KickStart/RoundA/workout.py
numTest = int(input().strip())
for testCase in range(1, numTest + 1):
  print("Case #", end = '')
  print(testCase, end = '')
  print(": ", end = '')
  n, k = map(int, input().strip().split())
  a = list(map(int, input().strip().split()))
  l = 1
  r = 1000000007
  ans = -1
  while l <= r:
    mid = (l + r) // 2
    numbAdd = 0
    for i in range(0, n - 1):
      diff = a[i + 1] - a[i]
      numbAdd += diff // mid
      if diff % mid == 0:
        numbAdd -= 1
    if numbAdd <= k:
      ans = mid
      r = mid - 1
    else:
      l = mid + 1
  print(ans)